### ISSUE TITLE
Discussion on batch fetch startegy

### ISSUE THREAD

**Issue by MichaelLiang12** [VisualComputingInstitute/triplet-reid#28]
In your TF codes, you first `shuffle` person ids and then repeat them forever. In training, you choose `batch_p` from the dataset according to the queue. For one person, you randomly choose `batch_k` examples each time. Am I right? For this situation, I have two questions:

1) The order of person ids is repeated. It means that each person will only be compared within `batch_p=25` people around them. You know, some person are easier to identify. You just maximize the margins in this small group.

2) How about choosing examples of one person in a `repeated way` ? It enables every example to be trained repeated. I know  that `randomly choose first batch_k` is theoretically OK. What about their differences on performance? Are them totally equivalent?

By the way, based on your codes, I try to implement a ResNet-50 fine-tuning baseline (just modify the last FC layer) for image classification (on CUB-200-2011). For testing, I feed data to the model in a normal ordered way. But the test accuracy (for classification) is only 22%. (It is supposed to be abound 81%). The training accuracy rises to 100% and the loss drops to 0.03 in 5000 iterations (about 100 epochs). Though training accuracy is of nonsense in this situation. What might be wrong? Is is due to the sampling strategy? Thank you.
![1](https://user-images.githubusercontent.com/17873150/38241913-059a31ca-3766-11e8-8626-c8bf69071f48.png) 
![1](https://user-images.githubusercontent.com/17873150/38241987-377adfd2-3766-11e8-92a9-66f90920f7cf.png)

**Comment by lucasb-eyer** [comment]
Thanks for your comments!

1. No, I believe your understanding of `tf.data` is incorrect; I recommend looking at [the documentation](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#shuffle). Specifically, there is the `reshuffle_each_iteration` flag which is enabled by default. Alternatively, you could just try it out in a small example, which is what we did back when we implemented it and the documentation was very bad :smile: 

2. Sorry, I don't understand what you mean, maybe you are right. Could you give an example of the sampling you are thinking of?

3. re CUB-200-2011: please keep issues focused and discuss separate points in separate issues. You actually [opened one for this already](https://github.com/VisualComputingInstitute/triplet-reid/issues/26). I don't think this has anything to do with the sampling, but rather what you describe is classic overfitting of a huge model on a tiny dataset, which can be combated in many different ways, but in general is hard to do.

**Comment by lucasb-eyer** [comment]
1. No worries, just always checking the docs is a good idea :wink: 
2. Oh, I think I see what you mean now. It's actually not a good idea in Market1501 to ensure all images are seen, because some PIDs have a huge number of images and many have just very few. IIRC the first two or so PIDs are actually the dataset authors and seen many, many times. That would bias the network a lot towards them, and lose importance of many "rare" PIDs. We actually did something similar to that in the very first experiments (in Theano) and it was worse maybe 1-2% mAP.
3. I see why you mention it in this issue now. Although I'm still somewhat confused, you are talking about softmax-classification now, right? I don't have experience with softmax-classification and this batching strategy, though what you report does sound wrong either way. Probably there is a mistake in the batch-loss computation? You could compute some other batch-level statistics, like the batch-level classification accuracy. But definitely write unit-tests for these things.
